Ensure light collections are dirtied on creation

aloysbaillet opened the pull request:
### Description of Change(s)

We found that when undoing the deletion of a light with an active collection, the re-creation of that light was not dirtying the content of the collection.

It is unfortunately quite tricky to provide a test for this use case, but we have one internally in Omniverse...
Our test goes like this:
* create 2 cubes
* create a light
* set the light's lightLink collection to refer to cube2
* check in viewport with a light-collection-enabled renderer that things are OK
* delete the light
* undo that deletion
* check in viewport with a light-collection-enabled renderer that things are OK again (it wasn't without this fix)

aloysbaillet committed:
Ensure light collections are dirtied on creation

spitzak commented:
We ran into this as well, but it was pretty simple to have the renderDelegate update things due to the creation of the light and the calling of Sync() on it, so I am not absolutely certain this dirtying is needed

aloysbaillet commented:
Thanks Bill, I'm still discovering the Hydra world... but in our render delegate the Light::Sync was indeed called, but each linked geometry rprim wasn't and we need them called to re-register them with the light link. I'm curious how you made it work?

spitzak commented:
This is lame but in Light::Sync() it checks if the categories changed and
if so it calls
sceneDelegate->GetRenderIndex().GetChangeTracker().MarkAllRprimsDirty(pxr::HdChangeTracker::DirtyCategories);
which calls Sync() on all the Geometry.


On Tue, Oct 12, 2021 at 4:34 PM Aloys Baillet ***@***.***>
wrote:

> Thanks Bill, I'm still discovering the Hydra world... but in our render
> delegate the Light::Sync was indeed called, but each linked geometry rprim
> wasn't and we need them called to re-register them with the light link. I'm
> curious how you made it work?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/PixarAnimationStudios/USD/pull/1653#issuecomment-941744171>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJMCEOOMZ3JDAUF377LBYTUGTAY5ANCNFSM5FZXXWVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

aloysbaillet commented:
Fair enough :)
I wanted to avoid marking all prims dirty though, the proposed change only marks the collection members dirty...

jilliene commented:
Filed as internal issue #USD-6961

tcauchois commented:
Hey Aloys,

Like I mentioned on your other PR, I'm a bit hesitant to have special case code in delegate.cpp since it becomes hard to maintain.  There's also a few bits you're missing.  This will be a bit involved to fix but I'll try to lay out the pieces and answer any questions.

First up, we have an object UsdImaging_CollectionCache that actually answers all of the queries, both GetLightParamValue(lightLink) and GetCategories(rprim).  Here's the class: https://github.com/PixarAnimationStudios/USD/blob/release/pxr/usdImaging/usdImaging/collectionCache.h ... and here's where we populate it: https://github.com/PixarAnimationStudios/USD/blob/release/pxr/usdImaging/usdImaging/lightAdapter.cpp#L131 ...  In order to keep the collection cache up to date with any changes, I'd recommend updating lightAdapter.cpp and lightFilterAdapter.cpp to call UpdateCollection in TrackVariability (which is already there); call UpdateCollection in ProcessPropertyChange if the collection changes (and return HdLight::DirtyParams); and call RemoveCollection in _RemovePrim.  This way the results of any calls into the collection cache will always be up-to-date.

The next piece you need is hydra dirty bit tracking.  Ideally, you could do this when you call UpdateCollection; the collection cache removes a bunch of gprims from the collection list, and adds a bunch of other ones, and you could just produce a list of paths to mark with DirtyCategories.  However, TrackVariability (for example) is called multi-threaded and can't call the hydra change-tracker, which isn't threadsafe.  And neither TrackVariability nor ProcessPropertyChange have a UsdImagingIndexProxy, which is the class needed to call MarkDirty.

The way out of this that I see is for the collection cache to keep a dirty gprim list.  Every time you call RemoveCollection or UpdateCollection, add hydra paths to an SdfPathSet (or Vector) on the collection cache itself.  At the end of UsdImagingDelegate::ApplyPendingUpdates(), which is our function that applies USD edits to imaging, loop through this set calling MarkDirty(DirtyCategories); I'm imagining putting that loop here https://github.com/PixarAnimationStudios/USD/blob/release/pxr/usdImaging/usdImaging/delegate.cpp#L1040 ... right after _ExecuteWorkForVariabilityUpdate.  Or the other option, of course, is Bill's approach of marking all rprims dirty with DirtyCategories whenever the collections change.  That's faster if your light collections tend to be big.  (Note that the trivial collection of "/" avoids all of these codepaths entirely).

That should do a much better job of handling the category change-tracking.

Hopefully this makes sense!  Any work to fix this up is super appreciated, and I'm happy to answer whatever questions you have about the code.

Thanks,
Tom

aloysbaillet commented:
Thanks Tom, I'll look into this!

aloysbaillet committed:
Added dirty tracking in CollectionCache

tcauchois commented on pxr/usdImaging/usdImaging/delegate.cpp:
```diff
@@ -1348,10 +1348,35 @@ UsdImagingDelegate::_ResyncUsdPrim(SdfPath const& usdPath,
+            };
+            UsdLuxLightAPI light(usdPrim);
+            UsdCollectionAPI lightLink{ light.GetLightLinkCollectionAPI() };
+            _MarkLinkPathsDirty(lightLink.ComputeIncludedPaths(lightLink.ComputeMembershipQuery(), _stage));
```
I think the "Compute" here won't go through the collection cache, right?  Is there a way to do that?

aloysbaillet replied:
This code is now deleted

tcauchois commented on pxr/usdImaging/usdImaging/delegate.cpp:
```diff
@@ -1082,6 +1082,39 @@ UsdImagingDelegate::ApplyPendingUpdates()
+        TF_FOR_ALL(it, _hdPrimInfoMap) {
+            UsdImagingPrimAdapterSharedPtr& adapter = it->second.adapter;
+            TF_DEBUG(USDIMAGING_CHANGES).Msg("[Update]: invalidate collection member prim (all paths dirty) %s\n", it->first.GetText());
+            adapter->MarkDirty(it->second.usdPrim, it->first,
```
We should only do this for gprims.  The standard way to do this is to add a "MarkCollectionsDirty" on prim adapter that does nothing, and an override on gprim adapter that calls MarkDirty(DirtyCategories).

"DirtyCollections" doesn't mean anything on other types of prim and will invalidate a random attribute.

tcauchois commented on pxr/usdImaging/usdImaging/lightAdapter.cpp:
```diff
@@ -177,6 +216,9 @@ UsdImagingLightAdapter::ProcessPropertyChange(UsdPrim const& prim,
         return HdLight::DirtyBits::DirtyTransform;
     }
+
+    _UpdateCollectionsChanged(prim, cachePath);
```
We'd want to guard this in "if (propertyName == lightLinkAttrToken)" or something, not sure what the right token is offhand.

aloysbaillet replied:
Indeed! Would this be OK? Some string parsing as the attributes can be numerous and are all under the collection:lightLink/shadowLink prefix:
```
    if (TfStringStartsWith(propertyName.GetString(), UsdImagingTokens->collectionShadowLink.GetString()) || 
        TfStringStartsWith(propertyName.GetString(), UsdImagingTokens->collectionLightLink.GetString()))
    {
        _UpdateCollectionsChanged(prim, cachePath);
    }
```

tcauchois replied:
Yeah, that would be great!

tcauchois commented on pxr/usdImaging/usdImaging/lightAdapter.cpp:
```diff
@@ -127,11 +163,14 @@ UsdImagingLightAdapter::TrackVariability(UsdPrim const& prim,
+        {
+            *timeVaryingBits |= HdLight::DirtyBits::DirtyCollection;
+        }
+        else
```
Can skip the else here.

tcauchois replied:
Actually, can just call _UpdateCollectionsChanged.  I don't think collections are allowed to be time-varying?  And if you did want to check that, instead of seeing if the collection has new contents, you'd want to ask whether the attributes are time varying.

aloysbaillet replied:
I removed the call altogether as the attributes cannot be animated... I'm guessing that if timesamples are added on the includeRoot bool that could cause some issues though?

tcauchois commented on pxr/usdImaging/usdImaging/lightAdapter.cpp:
```diff
@@ -80,6 +80,42 @@ UsdImagingLightAdapter::_RemovePrim(SdfPath const& cachePath,
+    };
+    UsdLuxLight light(prim);
+    const size_t newLightCollectionHash = getCollectionHash(light.GetLightLinkCollectionAPI());
+    const size_t newShadowCollectionHash = getCollectionHash(light.GetShadowLinkCollectionAPI());
```
We might have gone over this already, but why are we hashing the collections?  Can we not have "UpdateCollection" tell us whether the membership has changed?

aloysbaillet replied:
Indeed! Would this signature be OK?
```
TfToken
UsdImaging_CollectionCache::UpdateCollection(UsdCollectionAPI const& c, bool* changed)
```

aloysbaillet replied:
Actually I'll just return a bool instead of the TfToken as the return value isn't needed anymore...

tcauchois commented on pxr/usdImaging/usdImaging/lightAdapter.cpp:
```diff
@@ -80,6 +80,42 @@ UsdImagingLightAdapter::_RemovePrim(SdfPath const& cachePath,
+    SdfPath lightLinkPath = cachePath.AppendProperty(UsdImagingTokens->collectionLightLink);
+    collectionCache.RemoveCollection(GetDelegate()->GetStage(), lightLinkPath);
+    SdfPath shadowLinkPath = cachePath.AppendProperty(UsdImagingTokens->collectionShadowLink);
+    collectionCache.RemoveCollection(GetDelegate()->GetStage(), shadowLinkPath);
```
Might want to remove _collectionHashes[cachePath] here...

aloysbaillet replied:
Good point, but not needed anymore as I'm not storing the hashes any longer!

tcauchois commented on pxr/usdImaging/usdImaging/collectionCache.cpp:
```diff
@@ -78,15 +97,16 @@ UsdImaging_CollectionCache::UpdateCollection(UsdCollectionAPI const& c)
     _idForPath[path] = id;
 
+    _MarkCollectionContentDirty(stage, query);
+    
```
I feel like there's some missing bits here...  If _queryForId[id] before you update it != query, also _MarkContentDirty(stage, _queryForId[id])?

aloysbaillet replied:
The `RemoveCollection` call at the beginning of `UpdateCollection` will mark the previous collection content dirty, wouldn't this cover it?

tcauchois commented on pxr/usdImaging/usdImaging/delegate.h:
```diff
@@ -520,6 +520,12 @@ class UsdImagingDelegate : public HdSceneDelegate, public TfWeakBase {
     USDIMAGING_API
     bool IsInInvisedPaths(const SdfPath &usdPath) const;
 
+    USDIMAGING_API
```
I'd add this as protected API on primAdapter.cpp instead, if that's ok.

tcauchois commented on pxr/usdImaging/usdImaging/collectionCache.cpp:
```diff
@@ -38,10 +38,29 @@ _IsQueryTrivial(UsdCollectionAPI::MembershipQuery const& query)
+    {
+        // If there are any exlusion we have to consider all previously excluded paths dirty
+        // As there is no API to query excluded paths we mark all prims in the Stage dirty
+        _allPathsDirty = true;
```
Even if the query has exclude paths, I'd still expect UsdComputeIncludedPathsFromCollection to come up with the right list.  That should return the set of paths that pass both the include & exclude rules, right?

aloysbaillet replied:
Indeed, but prims that were previously excluded from the collection would not be returned in the include set, and these also need to be marked dirty...

aloysbaillet replied:
Actually as the RemoveCollection call will catch all paths from the outgoing collection, the UpdateCollection should be able to safely ignore the excluded paths... Will test that further!

aloysbaillet committed:
Cleaner and simpler collection dirty tracking

aloysbaillet committed:
Cleaner and more reliable collection dirty

aloysbaillet commented:
I've cleaned and simplified the code further, also fixed some formatting inconsistencies.
The `UsdImagingLightAdapter::ProcessPropertyChange` method now returns `HdLight::DirtyBits::DirtyCollection` when it detects a collection change (using hashes for comparison, but without having to store them).

I'm still having one refresh issue that I'm chasing but I'm confident it is a problem in our render delegate (I had to change it quite a lot to handle `HdLight::DirtyCollection` properly).
I'm quite confident the prman render delegate will need updating to take advantage of the now-working DirtyCollection flag, but I'm not sure if I'll be able to do that (I think I would need a license first ;) ).

aloysbaillet committed:
Missing method declarations

aloysbaillet committed:
Fixed compile error

aloysbaillet commented:
I've fixed a compilation error (`UsdLuxLight`->`UsdLuxLightAPI` as I was testing on an older USD version sorry!) and @sirpalee kindly tested these changes with success using the Arnold render delegate.

tcauchois commented on pxr/usdImaging/usdImaging/lightAdapter.cpp:
```diff
@@ -71,15 +71,50 @@ UsdImagingLightAdapter::Populate(UsdPrim const& prim,
+}
+
+bool
+UsdImagingLightAdapter::_UpdateCollectionsChanged(UsdPrim const& prim) const {
```
Probably just a merge issue, but _UpdateCollectionsChanged shows up twice in this file (seemingly with the same implementation).

aloysbaillet replied:
oh wow, sorry about this!

tcauchois commented on pxr/usdImaging/usdImaging/lightAdapter.cpp:
```diff
@@ -71,15 +71,50 @@ UsdImagingLightAdapter::Populate(UsdPrim const& prim,
+}
+
+bool
+UsdImagingLightAdapter::_UpdateCollectionsChanged(UsdPrim const& prim) const {
```
Probably just a merge issue, but _UpdateCollectionsChanged shows up twice in this file (seemingly with the same implementation).

aloysbaillet committed:
Removed duplicated method

tcauchois commented on pxr/usdImaging/usdImaging/lightAdapter.cpp:
```diff
@@ -71,15 +71,50 @@ UsdImagingLightAdapter::Populate(UsdPrim const& prim,
 }
 
 void
-UsdImagingLightAdapter::_RemovePrim(SdfPath const& cachePath,
```
Did you mean to remove this?  It's still marked overridden in the header, so it's causing build issues.

aloysbaillet replied:
Not really sorry about this, the USD version I've been testing with didn't have that method implemented!

tcauchois commented on pxr/usdImaging/usdImaging/lightAdapter.cpp:
```diff
@@ -71,15 +71,50 @@ UsdImagingLightAdapter::Populate(UsdPrim const& prim,
 }
 
 void
-UsdImagingLightAdapter::_RemovePrim(SdfPath const& cachePath,
```
Did you mean to remove this?  It's still marked overridden in the header, so it's causing build issues.

aloysbaillet committed:
Fixed another build issue
